les/clientpool.go

```diff
@@ -509,65 +622,108 @@ func (f *clientPool) getPosBalance(id enode.ID) *posBalance {
-// is updated to amount. This method also allows removing positive balance.
-func (f *clientPool) addBalance(id enode.ID, amount uint64, setTotal bool) {
-	f.lock.Lock()
-	defer f.lock.Unlock()
+// getPosBalanceIDs returns a lexicographically ordered list of IDs from accounts
+// with a positive balance
+func (f *clientPool) getPosBalanceIDs(start, stop enode.ID, maxCount int) (result []enode.ID) {
+	if maxCount <= 0 {
+		return
+	}
+	it := f.db.NewIteratorWithStart(append(clientBalanceDbKey, start[:]...))
+	defer it.Release()
+	for i := len(stop[:]) - 1; i >= 0; i-- {
+		stop[i]--
+		if stop[i] != 255 {
+			break
+		}
+	}
+	stopKey := append(clientBalanceDbKey, stop[:]...)
+
+	for it.Next() {
+		var id enode.ID
+		if len(it.Key()) != len(id[:])+len(clientBalanceDbKey) || bytes.Compare(it.Key(), stopKey) == 1 {
+			return
+		}
+		copy(id[:], it.Key()[len(clientBalanceDbKey):])
+		result = append(result, id)
+		if len(result) == maxCount {
+			return
+		}
+	}
+	return
+}
 
+// updateBalance updates the balance of a client (either overwrites it or adds to it).
+// It also updates the balance meta info string.
+func (f *clientPool) updateBalance(id enode.ID, amount int64, add bool, meta string) error {
 	pb := f.getPosBalance(id)
 	c := f.connectedMap[id]
 	var negBalance uint64
 	if c != nil {
 		pb.value, negBalance = c.balanceTracker.getBalance(f.clock.Now())
 	}
-	if setTotal {
-		if pb.value+amount > pb.lastTotal {
-			pb.value += amount - pb.lastTotal
+	if add {
+		if amount > 0 {
+			if amount > maxBalance || pb.value > maxBalance-uint64(amount) {
+				return errBalanceOverflow
+			}
+			pb.value += uint64(amount)
 		} else {
-			pb.value = 0
+			if uint64(-amount) > pb.value {
+				pb.value = 0
+			} else {
+				pb.value -= uint64(-amount)
+			}
 		}
-		pb.lastTotal = amount
 	} else {
-		pb.value += amount
-		pb.lastTotal += amount
+		if amount > maxBalance {
+			return errBalanceOverflow
+		}
+		if amount < 0 {
+			amount = 0
+		}
+		pb.value = uint64(amount)
 	}
+	pb.meta = meta
 	f.storePosBalance(pb)
 	if c != nil {
 		c.balanceTracker.setBalance(pb.value, negBalance)
 		if !c.priority && pb.value > 0 {
 			c.priority = true
+			if c.priority {
+				f.priorityConnected += c.capacity
+			}
 			c.balanceTracker.addCallback(balanceCallbackZero, 0, func() { f.balanceExhausted(id) })
 		}
+		c.balanceMetaInfo = meta
 	}
+	return nil
 }
 
 // posBalance represents a recently accessed positive balance entry
 type posBalance struct {
-	id                           enode.ID
-	value, lastStored, lastTotal uint64
-	queueIndex                   int // position in posBalanceQueue
+	id                enode.ID
+	value, lastStored uint64
+	meta              string
+	queueIndex        int // position in posBalanceQueue
 }
 
 // EncodeRLP implements rlp.Encoder
 func (e *posBalance) EncodeRLP(w io.Writer) error {
-	return rlp.Encode(w, []interface{}{e.value, e.lastTotal})
+	return rlp.Encode(w, []interface{}{e.value, e.meta})
```

Comment:
The database schema change will definitely introduce some troubles. e.g. we should consider backward-compatibility

Reply:
What do you mean? This code is not used yet by anyone so there is no compatibility issue. The `meta` field should be flexible enough so we don't need to add more application specific fields. The script using this API can use the `meta` field for whatever it needs, for example it can store the id of the last processed incoming tx to make sure that everything is added exactly once.